// n8n/dist/runners/runner-ws-server.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskRunnerWsServer = void 0;
const typedi_1 = require("typedi");
const logger_service_1 = require("../logging/logger.service");
const default_task_runner_disconnect_analyzer_1 = require("./default-task-runner-disconnect-analyzer");
const task_broker_service_1 = require("./task-broker.service");
function heartbeat() {
    this.isAlive = true;
}
let TaskRunnerWsServer = class TaskRunnerWsServer {
    constructor(logger, taskBroker, disconnectAnalyzer) {
        this.logger = logger;
        this.taskBroker = taskBroker;
        this.disconnectAnalyzer = disconnectAnalyzer;
        this.runnerConnections = new Map();
    }
    setDisconnectAnalyzer(disconnectAnalyzer) {
        this.disconnectAnalyzer = disconnectAnalyzer;
    }
    getDisconnectAnalyzer() {
        return this.disconnectAnalyzer;
    }
    sendMessage(id, message) {
        this.runnerConnections.get(id)?.send(JSON.stringify(message));
    }
    add(id, connection) {
        connection.isAlive = true;
        connection.on('pong', heartbeat);
        let isConnected = false;
        const onMessage = async (data) => {
            try {
                const buffer = Array.isArray(data) ? Buffer.concat(data) : Buffer.from(data);
                const message = JSON.parse(buffer.toString('utf8'));
                if (!isConnected && message.type !== 'runner:info') {
                    return;
                }
                else if (!isConnected && message.type === 'runner:info') {
                    await this.removeConnection(id);
                    isConnected = true;
                    this.runnerConnections.set(id, connection);
                    this.taskBroker.registerRunner({
                        id,
                        taskTypes: message.types,
                        lastSeen: new Date(),
                        name: message.name,
                    }, this.sendMessage.bind(this, id));
                    this.logger.info(`Runner "${message.name}" (${id}) has been registered`);
                    return;
                }
                void this.taskBroker.onRunnerMessage(id, message);
            }
            catch (error) {
                this.logger.error(`Couldn't parse message from runner "${id}"`, {
                    error: error,
                    id,
                    data,
                });
            }
        };
        connection.once('close', async () => {
            connection.off('pong', heartbeat);
            connection.off('message', onMessage);
            await this.removeConnection(id);
        });
        connection.on('message', onMessage);
        connection.send(JSON.stringify({ type: 'broker:inforequest' }));
    }
    async removeConnection(id) {
        const connection = this.runnerConnections.get(id);
        if (connection) {
            const disconnectReason = await this.disconnectAnalyzer.determineDisconnectReason(id);
            this.taskBroker.deregisterRunner(id, disconnectReason);
            connection.close();
            this.runnerConnections.delete(id);
        }
    }
    handleRequest(req, _res) {
        this.add(req.query.id, req.ws);
    }
};
exports.TaskRunnerWsServer = TaskRunnerWsServer;
exports.TaskRunnerWsServer = TaskRunnerWsServer = __decorate([
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [logger_service_1.Logger,
        task_broker_service_1.TaskBroker,
        default_task_runner_disconnect_analyzer_1.DefaultTaskRunnerDisconnectAnalyzer])
], TaskRunnerWsServer);
