// n8n/dist/src/databases/utils/transformers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sqlite = exports.objectRetriever = exports.lowerCaser = exports.idStringifier = void 0;
const config_1 = __importDefault(require("../../../config"));
exports.idStringifier = {
    from: (value) => (typeof value === 'number' ? value.toString() : value),
    to: (value) => (typeof value === 'string' ? Number(value) : value),
};
exports.lowerCaser = {
    from: (value) => value,
    to: (value) => (typeof value === 'string' ? value.toLowerCase() : value),
};
exports.objectRetriever = {
    to: (value) => value,
    from: (value) => typeof value === 'string' ? JSON.parse(value) : value,
};
const jsonColumn = {
    to: (value) => config_1.default.getEnv('database.type') === 'sqlite' ? JSON.stringify(value) : value,
    from: (value) => typeof value === 'string' ? JSON.parse(value) : value,
};
exports.sqlite = { jsonColumn };
