// n8n-editor-ui/dist/assets/ResourceListHeader-B2sxe6HT.js
import { d as defineComponent, c as openBlock, h as createElementBlock, j as createBaseVNode, i as createVNode, A as unref, iM as N8nIcon, n as normalizeClass, w as withCtx, q as renderSlot, i3 as N8nHeading, e as createBlock, iN as N8nText, f as createCommentVNode, _ as _export_sfc } from "./index-7YFQEQmt.js";
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "ResourceListHeader",
  props: {
    icon: {}
  },
  setup(__props) {
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", {
        class: normalizeClass([_ctx.$style.workflowHeader])
      }, [
        createBaseVNode("div", {
          class: normalizeClass([_ctx.$style.icon])
        }, [
          createVNode(unref(N8nIcon), {
            icon: _ctx.icon,
            color: "text-light"
          }, null, 8, ["icon"])
        ], 2),
        createBaseVNode("div", null, [
          createVNode(unref(N8nHeading), {
            bold: "",
            tag: "h2",
            size: "xlarge"
          }, {
            default: withCtx(() => [
              renderSlot(_ctx.$slots, "title")
            ]),
            _: 3
          }),
          _ctx.$slots.subtitle ? (openBlock(), createBlock(unref(N8nText), {
            key: 0,
            size: "small",
            color: "text-light"
          }, {
            default: withCtx(() => [
              renderSlot(_ctx.$slots, "subtitle")
            ]),
            _: 3
          })) : createCommentVNode("", true)
        ]),
        _ctx.$slots.actions ? (openBlock(), createElementBlock("div", {
          key: 0,
          class: normalizeClass([_ctx.$style.actions])
        }, [
          renderSlot(_ctx.$slots, "actions")
        ], 2)) : createCommentVNode("", true)
      ], 2);
    };
  }
});
const workflowHeader = "_workflowHeader_3z0bo_1";
const icon = "_icon_3z0bo_9";
const actions = "_actions_3z0bo_15";
const style0 = {
  workflowHeader,
  icon,
  actions
};
const cssModules = {
  "$style": style0
};
const ResourceListHeader = /* @__PURE__ */ _export_sfc(_sfc_main, [["__cssModules", cssModules]]);
export {
  ResourceListHeader as R
};
